pyro/contrib/oed/eig.py

```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
+
+    [1] Kleinegesse, Steven, and Michael Gutmann. "Efficient Bayesian Experimental Design for Implicit Models."
+    arXiv preprint arXiv:1810.09912 (2018).
+
+    :param function model: A pyro model accepting `design` as only argument.
+    :param torch.Tensor design: Tensor representation of design
+    :param list observation_labels: A subset of the sample sites
+        present in `model`. These sites are regarded as future observations
+        and other sites are regarded as latent variables over which a
+        posterior is to be inferred.
+    :param list target_labels: A subset of the sample sites over which the posterior
+        entropy is to be measured.
+    :param int num_y_samples: Number of samples to take in :math:`y` for each :math:`\\theta`.
+    :param: int num_theta_samples: Number of initial samples in :math:`\\theta` to take. The likelihood ratio
+                                   is estimated by LFIRE for each sample.
+    :param int num_steps: Number of optimisation steps.
+    :param function classifier: a Pytorch or Pyro classifier used to distinguish between samples of :math:`y` under
+                                :math:`p(y|d)` and samples under :math:`p(y|\\theta,d)` for some :math:`\\theta`.
+    :param pyro.optim.Optim optim: Optimiser to use.
+    :param bool return_history: If `True`, also returns a tensor giving the loss function
+        at each step of the optimisation.
+    :param torch.Tensor final_design: The final design tensor to evaluate at. If `None`, uses
+        `design`.
+    :param int final_num_samples: The number of samples to use at the final evaluation, If `None,
+        uses `num_samples`.
+    :return: EIG estimate, optionally includes full optimisation history
+    :rtype: `torch.Tensor` or `tuple`
+    """
+    if isinstance(observation_labels, str):
+        observation_labels = [observation_labels]
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+
+    # Take N samples of the model
+    expanded_design = lexpand(design, num_theta_samples)
+    trace = poutine.trace(model).get_trace(expanded_design)
+
+    theta_dict = {l: trace.nodes[l]["value"] for l in target_labels}
+    cond_model = pyro.condition(model, data=theta_dict)
+
+    loss = _lfire_loss(model, cond_model, classifier, observation_labels, target_labels)
+    out = opt_eig_ape_loss(expanded_design, loss, num_y_samples, num_steps, optim, return_history,
+                           final_design, final_num_samples)
+    if return_history:
+        return out[0], out[1].sum(0) / num_theta_samples
+    else:
+        return out.sum(0) / num_theta_samples
+
+
+def vnmc_eig(model, design, observation_labels, target_labels,
+             num_samples, num_steps, guide, optim, return_history=False,
+             final_design=None, final_num_samples=None):
+    """Estimates the EIG using Variational Nested Monte Carlo (VNMC). The VNMC estimate is
```

Comment:
here and elsewhere you could you some foster et al references?

Reply:
I will add a Foster et al. reference to the four estimators mentioned in our paper